crates/pallet-feeds/src/lib.rs

```diff
@@ -122,32 +111,43 @@ mod pallet {
         pub fn put(
             origin: OriginFor<T>,
             feed_id: FeedId,
-            data: PutDataObject,
-            metadata: ObjectMetadata,
+            object: Object,
+            metaobject: Metaobject,
         ) -> DispatchResult {
             let who = ensure_signed(origin)?;
 
-            let object_size = data.len() as u64;
+            let object_size = object.len() as u64;
 
-            log::debug!("metadata: {:?}", metadata);
-            log::debug!("object_size: {:?}", object_size);
+            log::debug!("object_size: {}, metaobject: {:?}", object_size, metaobject);
 
             let current_feed_id = Self::current_feed_id();
-
             ensure!(current_feed_id >= feed_id, Error::<T>::UnknownFeedId);
 
-            Metadata::<T>::insert(feed_id, metadata.clone());
-
-            Totals::<T>::mutate(feed_id, |feed_totals| {
-                feed_totals.size += object_size;
-                feed_totals.count += 1;
+            Metaobjects::<T>::insert(feed_id, metaobject.clone());
+            Feeds::<T>::mutate(feed_id, |feed_info| {
+                feed_info.total_size += object_size;
+                feed_info.total_objects += 1;
             });
 
-            Self::deposit_event(Event::DataSubmitted(metadata, who, object_size));
+            Self::deposit_event(Event::ObjectSubmitted(who, metaobject, object_size));
 
             Ok(())
         }
     }
+
+    #[pallet::storage]
+    #[pallet::getter(fn metaobjects)]
+    pub(super) type Metaobjects<T: Config> =
+        StorageMap<_, Blake2_128Concat, FeedId, Metaobject, OptionQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn feeds)]
+    pub(super) type Feeds<T: Config> =
+        StorageMap<_, Blake2_128Concat, FeedId, FeedInfo, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn current_feed_id)]
+    pub(super) type CurrentFeedId<T: Config> = StorageValue<_, FeedId, ValueQuery>;
```

Comment:
Why are these moved? Usually storage items come before calls that act upon them.

Reply:
With a quick look, we can see a bunch of substrate pallets put `pallet::storage` before `pallet::call` in practice, but there are also quite a few of them that do the opposite, especially the **example** pallet, what's more, according to https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines, the store items(decl_storage) comes after the pallet calls(decl_module).

>3. reorganize pallet to have trait Config, decl_* macros, ValidateUnsigned, ProvideInherent, Origin all together in one file. Suggested order:
>
>- Config,
>- decl_module,
>- decl_event,
>- decl_error,
>- decl_storage,
>- origin,
>- validate_unsigned,
>- provide_inherent, so far it should compile and all be correct.

So, I'd say the initial idea is to put `pallet::storage` after `pallet::call`, even though they follow the recommended rule strictly in practice.

Reply:
If I understand correctly that is suggested order for reorganization, not necessarily order of placing things.

Is the fact that some pallets have it the other way around the only reason to move it?

If you read a book, you first meet definitions, people, some backround information, prerequisites for the rest of the content. Same here, the code should be possible to read as a story or an article, you should be able to explain what is happening in the code as you go top to bottom with as little jumps back and forth and mostly without repeating the code itself literally. You need to know the data you work with before you use it. The closer code is to that ideal, the better I it is in my opinion.

Reply:
>If I understand correctly that is suggested order for reorganization, not necessarily order of placing things.

That implicitly indicates the recommended order of pallet elements to my understanding. And the example pallet that is used for demonstration adopts it as well.

>Is the fact that some pallets have it the other way around the only reason to move it?

Not really. It's merely one of my considerations, all the factors combined contribute to my final view.

The code and data are basically two essentials at a high level, I don't have a strong opinion that which one is more important or should appear first, without any of them, we can't make a real program. The code means the business logic, the data is more about the implementation details and serves the business logic IMHO. In a perfect world, we might just describe what we want and let AI think about the data structure, and implement all the logic. But we're not in a perfect world, so no 100% answer to this. 1 plus 9 equals 10, 9 plus 1 equals 10 too, again, no strong opinion on this, just thought that might be the recommended way.

Reply:
I disagree that data is implementation detail. Data is the primary thing we care about. Correctly modeled data structure make it easier to deal with. In fact rich type system in Rust allows to model data in ways that some other languages do not support. At the start and the end of the program all we care is data, it needs to be transformed somehow, sure, but it is all about data one way or another. Business logic is an implementation detail how data is transformed from one shape into another, not the other way around.

Reply:
That's an open question and mostly it depends on where you stand. It's true if from the view of the developer. From the user perspective, the underlying data structure is not that important, either two maps or a double map is fine, just keep my balances correctly and let me trade as I wish.